runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/DispatchAgent.java

```diff
@@ -1492,10 +1549,43 @@ private Target supplyTarget(
         return targetsByIndex.computeIfAbsent(index, newTarget);
     }
 
+    private LongConsumer supplyMetricWriter(
+        Metric.Kind kind,
+        long bindingId,
+        long metricId)
+    {
+        return metricWriterSuppliers.get(kind).apply(bindingId, metricId);
+    }
+
+    // Each binding has a list of metrics configured and the BindingRegistry has the chained up MetricHandler that takes
+    // care of recording all the configured metrics. This function returns the MetricHandler that belongs to the given binding.
+    private MetricHandler supplyMetricRecorder(
+        MetricHandlerKind kind,
+        long bindingId)
+    {
+        BindingRegistry binding = configuration.resolveBinding(bindingId);
+        MetricHandler result;
+        if (binding == null)
+        {
+            result = MetricHandler.NO_OP;
+        }
+        else if (kind == ORIGIN)
+        {
+            result = binding.originMetricHandler();
+        }
+        else
+        {
+            assert kind == ROUTED;
+            result = binding.routedMetricHandler();
+        }
+        return result;
```

Comment:
This code is going to execute for every frame we read and write, so better to split into `supplyOriginMetricRecorder` and `supplyRoutedMetricRecorder` to eliminate the repeated conditional logic.

```
    private MetricHandler supplyOriginMetricRecorder(
        long bindingId)
    {
        BindingRegistry binding = configuration.resolveBinding(bindingId);
        return binding != null ? binding.originMetricHandler() : MetricHandler.NO_OP;
    }

    private MetricHandler supplyRoutedMetricRecorder(
        long bindingId)
    {
        BindingRegistry binding = configuration.resolveBinding(bindingId);
        return binding != null ? binding.routedMetricHandler() : MetricHandler.NO_OP;
    }
```

Reply:
fixed